src/renderer/atlas/AtlasEngine.h

```diff
@@ -0,0 +1,761 @@
+        //
+        // But std::string employs an optimization called "small string optimization" (SSO).
+        // To simplify things it could be explained as:
+        // If the string capacity is small, then the characters are stored inside the "data"
+        // pointer and you make sure to set the lowest bit in the pointer one way or another.
+        // Heap allocations are always aligned by at least 4-8 bytes on any platform.
+        // If the address of the "data" pointer is not even you know data is stored inline.
+        template<typename T>
+        union SmallObjectOptimizer
+        {
+            static_assert(std::is_trivially_copyable_v<T>);
+            static_assert(std::has_unique_object_representations_v<T>);
+
+            T* allocated = nullptr;
+            T inlined;
+
+            constexpr SmallObjectOptimizer() = default;
+
+            SmallObjectOptimizer(const SmallObjectOptimizer& other)
+            {
+                const auto otherData = other.data();
+                const auto otherSize = other.size();
+                const auto data = initialize(otherSize);
+                memcpy(data, otherData, otherSize);
+            }
+
+            SmallObjectOptimizer& operator=(const SmallObjectOptimizer& other)
+            {
+                if (this != &other)
+                {
+                    delete this;
+                    new (this) SmallObjectOptimizer(other);
+                }
+                return &this;
+            }
+
+            SmallObjectOptimizer(SmallObjectOptimizer&& other) noexcept
+            {
+                memcpy(this, &other, std::max(sizeof(allocated), sizeof(inlined)));
+                other.allocated = nullptr;
+            }
+
+            SmallObjectOptimizer& operator=(SmallObjectOptimizer&& other) noexcept
+            {
+                return *new (this) SmallObjectOptimizer(other);
+            }
+
+            ~SmallObjectOptimizer()
+            {
+                if (!is_inline())
+                {
+#pragma warning(suppress : 26408) // Avoid malloc() and free(), prefer the nothrow version of new with delete (r.10).
+                    free(allocated);
+                }
+            }
+
+            T* initialize(size_t byteSize)
+            {
+                if (would_inline(byteSize))
+                {
+                    return &inlined;
+                }
+
+#pragma warning(suppress : 26408) // Avoid malloc() and free(), prefer the nothrow version of new with delete (r.10).
+                allocated = THROW_IF_NULL_ALLOC(static_cast<T*>(malloc(byteSize)));
+                return allocated;
+            }
+
+            constexpr bool would_inline(size_t byteSize) const noexcept
+            {
+                return byteSize <= sizeof(T);
+            }
+
+            bool is_inline() const noexcept
+            {
+                return (__builtin_bit_cast(uintptr_t, allocated) & 1) != 0;

```

Comment:
Sorry, I don't totally understand. Since this is a union, how do we ensure that the _inline value_ (which shares storage with `allocated`) doesn't have this bit set in this position?

Reply:
Okay, I had to think about how this works for a while. This will always hold a T, either internal or allocated, but sometimes that T will have been allocated with extra space at the end.
The T inside this union needs to be complicit - it has to have a member at the end that is effectively flexible but also has the appropriate amount of padding (if it wants).

I still don't understand how we ensure that the inlined value doesn't have this bit set.

Reply:
`SmallObjectOptimizer` is pretty much an implementation detail of `AtlasKey` and `AtlasValue` and the code doesn't make any sense whatsoever outside of that context.
The reason this works is because both `AtlasKeyData` as well as `AtlasValueData` coincidentially don't use the first bit for anything except for flagging whether inlined storage is used (for instance `CellFlags::Inlined`).

This code doesn't have to be super duper good, because I'm planning to make it a lot more robust and simpler with my custom hashmap implementation, which is required in order to get the LRU behavior we need. It will simultaneously allow us write simpler code here, as we can choose ourselves how to manage our linking behavior. For instance we don't need to allocate `AtlasValue` on the heap, if we simply use the linked list of the LRU itself to chain multiple atlas positions together.
In either case I expect this code to be rewritten and only the idea to remain, but in an improved form.